amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,554 @@
+                        props["div_min"], props["div_max"], props["div_step"], self.osc_div))
+
+                clk_i = Signal()
+                if osc == "OSCZ":
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             i_OSCEN=Const(1),
+                                             o_OSCOUT=clk_i)
+                elif osc == "OSCO":
+                    # TODO: Make use of regulator configurable
+                    m.submodules += Instance(osc,
+                                             p_REGULATOR_EN=Const(1),
+                                             p_FREQ_DIV=self.osc_div,
+                                             i_OSCEN=Const(1),
+                                             o_OSCOUT=clk_i)
+                elif osc == "OSCF":
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             o_OSCOUT30M=None,
+                                             o_OSCOUT=clk_i)
+                else:
+                    m.submodules += Instance(osc,
+                                             p_FREQ_DIV=self.osc_div,
+                                             o_OSCOUT=clk_i)
+
+            else:
+                clk_i = self.request(self.default_clk).i
+
+            if self.default_rst is not None:
+                rst_i = self.request(self.default_rst).i
+            else:
+                rst_i = Const(0)
+
+            m.submodules.reset_sync = ResetSynchronizer(rst_i, domain="sync")
+            m.domains += ClockDomain("sync")
+            m.d.comb += ClockSignal("sync").eq(clk_i)
+            return m
+
+    def _get_xdr_buffer(self, m, pin, i_invert=False, o_invert=False):
+
+        def get_ireg(clk,d,q):
+            for bit in range(len(q)):
+                m.submodules += Instance("DFF",
+                    i_CLK = clk,
+                    i_D = d[bit],
+                    o_Q = q[bit],
+                )
+
+        def get_oreg(clk,d,q):
+            for bit in range(len(q)):
+                m.submodules += Instance("DFF",
+                    i_CLK = clk,
+                    i_D = d[bit],
+                    o_Q = q[bit]
+                )
+
+        def get_iddr(clk,d,q0,q1):
+            for bit in range(len(d)):
+                m.submodules += Instance("IDDR",
+                    i_CLK = clk,
+                    i_D = d[bit],
+                    o_Q0 = q0[bit],
+                    o_Q1 = q1[bit]
+                )
+
+        def get_oddr(clk,d0,d1,q):
+            for bit in range(len(q)):
+                m.submodules += Instance("ODDR",
+                    i_CLK = clk,
+                    i_D0 = d0[bit],
+                    i_D1 = d1[bit],
+                    o_Q0 = q[bit],
```

Comment:
I think that this isn't quite the right implementation. On Gowin, `ODDR` has two flops inside: one for the data input, and one for the tristate input. I think you have to put both `o` and `oe` through the `ODDR` primitive, and then connect `Q1` to the `t` output.

Reply:
Ah, yes! I checked the manual and found the description in section 4.3.1 of [UG289E](http://cdn.gowinsemi.com.cn/UG289E.pdf). Is this mainly to cope with the associated delay in the DDR path so that signals align properly at the input of the io/tristate buffer?

Reply:
Yep

Reply:
The manual claims that Q1 can also be left floating.

Reply:
Yes, if you don't need tristates. But we do have a case where you have a DDR tristate, don't we?

Reply:
Yes, that is when `pin.xdr == 2` and `pin.dir in ("oe", "io")`.

Reply:
Since the ODDR has two flops already, one doesn't need an extra one, and one condition for all of `"o", "oe" and "io"` should be sufficient?

Reply:
Yes, that was essentially my point here.